### PR TITLE
 Optimize memory allocation of LCS algorithm

### PR DESCRIPTION
**Pull request by tmat**
Consolidate allocations of many small int arrays to a few allocations of larger buffers.
Pool buffers.

Fixes https://devdiv.visualstudio.com/DevDiv/_queries/edit/947470/?triage=true

### CODE REVIEW

**Comment by tmat** [comment]
@ivanbasov @sharwell @heejaechang PTAL

**Comment by CyrusNajmabadi** [comment]
Dumb question.  We also have an LCS algorithm for doing the NavTo search.  There, we just always pooled the normal array buffers, with almost no logic surrounding it.  And it seemed to work out fine like that.  The net amount of memory kept around was tiny, and basically the pool was usable for pretty much 99.9% of cases.  It was simple and effective.  Can we not do the same here?

**Comment by tmat** [comment]
What sizes of sequences does the NavTo search deal with? The problem I see with having a lot of small arrays in a pool is that the pooled objects will likely get promoted to Gen2 and then the GC will have to deal with a lot of references to small objects.

**Comment by tmat** [comment]
@CyrusNajmabadi Can you point me to the LCS implementation in NavTo search. I can't find it :(

**Comment by CyrusNajmabadi** [comment]
>and then the GC will have to deal with a lot of references to small objects.

Well, if they're gen2, doesn't the GC not have to deal with them?   I found for navto that capping at 128 elements was fine.  But that was just domain specific, not a general recommendation on size.  For example, int he sql space, we cap at 4k.   Those arrays do make it to gen2 since they stay around effectively forever.  But we get near 99.99% resuse on them without allocating other arrays.  I think, for sqlite we ended up with basically around 40k total allocated for this pooled data, which seemed totally fine and nothing that needed to be optimized or have complex logic around.

**Comment by CyrusNajmabadi** [comment]
> @CyrusNajmabadi Can you point me to the LCS implementation in NavTo search. I can't find it :(

Sure!  It's here: https://github.com/dotnet/roslyn/blob/master/src/Workspaces/Core/Portable/Utilities/EditDistance.cs

Note, we just use the final distance (we don't actually need the edit script).  But we do use a two-dimensional 64x64 array (my bad on saying it was 128, faulty memory on my part).   So that's basically 8k for each pooled table.    I looked into the simplification where you only need the last two rows, but it turned out to just not be a measurable savings *at all*.

Basically, we end up only needing to pool up to generally 2x core-count for these since a task wakes up, does the check, then puts the table back in the pool.  I've never been quite sure why that's around 2x core count, instead of core-count.  But the memory usage has always been so tiny, i've never cared to deeply lookinto things.

**Comment by tmat** [comment]
> Well, if they're gen2, doesn't the GC not have to deal with them? 
They are not Gen2 right away. They first need to get promoted and then moved around etc. 

> 64x64 array is tiny in comparison to what EnC needs.

For EnC we are diffing syntax trees, which means the lengths of the sequences might be from a few to 10000 (e.g. for huge files). I saw a memory dump from a customer where the sequence was 5000 elements long, which means the algorithm will allocate arrays of all sizes in [3, 2*1+1, 2*2+1, ..., 2*5000+1]. Rather than have 1000s of arrays allocated I think it's better to have a few large buffers.

File: src/Workspaces/Core/Portable/Differencing/LongestCommonSubsequence.cs [hunk 1]
```diff
@@ -11,64 +12,202 @@ namespace Microsoft.CodeAnalysis.Differencing
...
+        /// Instead we allocate a large buffer serving as a a backing storage of a contiguous sequence of arrays 
+        /// corresponding to stack depths <see cref="MinDepth"/> to <see cref="MaxDepth"/>.
+        /// If more storage is needed we chain next large buffer to the previous one in a linked list.
+        /// 
+        /// We pool a few of these linked buffers on <see cref="VStack"/> to conserve allocations.
+        /// </remarks>
+        private sealed class VBuffer
         {
+            /// <summary>
+            /// The max stack depth backed by the fist buffer.
```
Comment (ivanbasov) [inline src/Workspaces/Core/Portable/Differencing/LongestCommonSubsequence.cs]: >fist [](start = 50, length = 4)

first

**Comment by CyrusNajmabadi** [comment]
Oh, for enc do you need subarrays allocated as you're doing work? i.e. is it not sufficient to just have one array allocated for al lthe work (which can then be returned when done to be used by the next enc case)?  Or do you need potentially a *lot* of arrays available during one effective "enc op"?

If the latter, then this makes a lot more sense to me.  Thanks!

**Comment by tmat** [comment]
We used to have the simpler implemantation that used a matrix like in NavTo. But we OOM'd on that 5000x5000x4 ~ 100MB case :(
So we optimized the algorithm to use O(N*(L1+L2)) memory rather than O(N^2), but that meant allocating a lot of small arrays.
EnC operates on a very different scale than NavTo.

**Comment by tmat** [comment]
See https://github.com/dotnet/roslyn/pull/37658/files#diff-d758865430a7a9136396a566e6d4ed91R345

**Comment by CyrusNajmabadi** [comment]
> We used to have the simpler implemantation that used a matrix like in NavTo.

Sure.  I'm not arguing to use a matrix *at all*.  That wasn't my point.  My argument was to pick a large array for the 1-2 rows you need to keep.  After all, say you even had the row be 20k.  Two rows at 20k pointers.  is just a tiny 300k of memory.  Something that would be lost in the noise of doing enc work i imagine.  

The only reason i brought up the matrix/row thing was to simply say that that *particular* optimization wasn't necessary for NavTo, not to indicate that the matrix solution was the way to go.  

Anyways, up to you on how to proceed.

**Comment by tmat** [comment]
> My argument was to pick a large array for the 1-2 rows you need to keep.

I don't understand. The EnC LCS algorithm works differently from the other algorithm. We need the full graph for constructing an edit script. I can't just use last 2 rows.

**Comment by CyrusNajmabadi** [comment]
> I don't understand. The EnC LCS algorithm works differently from the other algorithm. We need the full graph for constructing an edit script. I can't just use last 2 rows.

This is why i asked:

> is it not sufficient to just have one array allocated for al lthe work (which can then be returned when done to be used by the next enc case)? Or do you need potentially a lot of arrays available during one effective "enc op"?

Basically, from the PR and my existing knowledge of LCS algorithms I wasn't certain what your needs were.  If it was just the standard LCS algorithm that only needs two rows, then i thought you'd be fine just pooling those arrays and reusing them.  As is said, if your needs are different then:

> If the latter, then this makes a lot more sense to me. Thanks!

I can only go on my understanding and the information provided in response to questions :)   It looks like there's stuff about enc that wasn't apparent just from a discussion about this particular utility library.  As such, i can understand why further optimizations might be necessary. Thanks!
